src/systolic_array/unit_test/test_tensor_iterator.cpp

```diff
@@ -0,0 +1,149 @@
+#include <iostream>
+#include <vector>
+#include <memory>
+
+#include "aladdin/unit-test/catch.hpp"
+#include "systolic_array/tensor.h"
+
+using namespace systolic;
+
+// Compute the reference results for region-based iterator operation. It's slow
+// as it literally converts the region into single step movements, but it gives
+// us the correct results.
+std::vector<int> computeReferenceResults(const TensorIndexIterator& iter,
+                                         const std::vector<int>& region) {
+  std::unique_ptr<TensorIndexIterator> refIter(iter.clone());
+  int numSingleSteps = 0;
+  int stride = 1;
+  for (int i = region.size() - 1; i >= 0; i--) {
+    numSingleSteps += region[i] * stride;
+    stride *= iter.getDimSize(i);
+  }
+  for (int i = 0; i < numSingleSteps; i++)
+    ++(*refIter);
+  return refIter->getIndices();
+}
+
+void doRegionTest(TensorIndexIterator& iter, const std::vector<int>& region) {
+  std::vector<int> refResults(computeReferenceResults(iter, region));
+  iter += region;
+  REQUIRE(iter.getIndices() == refResults);
+}
+
+SCENARIO("Test tensor index iterator", "[tensor-index]") {
+  GIVEN("Test tensor shape without paddings.") {
+    TensorShape shape({ 4, 8, 8, 16 });
+    TensorIndexIterator iter(shape);
+    WHEN("Test single step advance.") {
+      THEN("Advance the iterator by 1 step.") {
+        ++iter;
+        REQUIRE(int(iter) == 1);
+      }
+      THEN("Advance the iterator by a random number of single steps.") {
+        for (int i = 0; i < 1234; i++)
+          ++iter;
+        REQUIRE(int(iter) == 1234);
+      }
+    }
+    WHEN("Test regin-based advance.") {
+      THEN("Advance the iterator by regions that don't go out of dimension "
```

Comment:
This overlaps with the previous question, but is this a reasonable input (exceeding some dimension bounds)?

Reply:
Yes, it is a valid input, even though it goes out of that dimension bound. For example, a tensor shape {1, 4, 4, 8} with the iterator at {0, 0, 0, 0}, and now the systolic array wants to advance the iterator at the W dimension (NHWC layout) by 8, i.e. a region of {0, 0, 8, 0} to advance, just because the accelerator has 8 fetchers, so after each fetcher finishes its own job, it advances the iterator by 8 to start the next job. In this case, it should advance the iterator to {0, 2, 0, 0}.

Reply:
Gotcha, thanks for the explanation.